### DISCUSSION TITLE
Listening for OS theme change

### DISCUSSION THREAD

**Discussion by alienator88** [wailsapp/wails#2035] [Q&A] [answered]
Hi everyone,

Quick rundown, I have an electron app I'm converting to use Wails. On the Electron side, there's a module that allows you to subscribe to system preferences changes using Apple's "NSDistributedNotificationCenter" API. 
Reference here: https://www.electronjs.org/docs/latest/api/system-preferences#systempreferencessubscribenotificationevent-callback-macos

With this I can listen for when the OS changes to Dark mode or Light mode and emit an event to have my CSS change up to match the colors better depending on the theme. 

Right now with Wails, if the OS theme changes, colors tend to disappear when you have white text with the Light theme and vice versa.

Is there a way in Wails to listen for this change yet? I tried playing around with the Appearance Options but couldn't get anything to output from there.
I do have a hacky workaround for now that execs some applescript which gets the dark mode setting when the JS side is up and runs, but it changes the colors right after the window displays and doesn't look great as you can see it flash and change colors really quick. 

Would rather make these changes during the startup somehow. I tried emitting an event during the Go startup function as well, but the JS side doesn't seem to be up and listening yet for the event when it gets sent so it never sees it. It's possible I'm doing something wrong as well.

Thank you!

**Comment by leaanthony** [comment, marked as answer]
Hi 👋 It exists on a branch so we could bring it in at some point. There's a nuance that on Windows you have an additional "high contrast" mode per theme. What do you think the API would look like? I definitely think you should open an enhancement ticket with the suggestions 👍

On a side note, we need to shore up system eventing.

**Comment by alienator88** [comment]
I was thinking something like the electron API, just not sure how feasible that would be for other platforms like Windows/Linux:
1. Have the app listening for the OS theme change via a subscription mechanism.
2. When it receives the change notification, user can then execute any function they'd want to alter the look of their application to match the current OS appearance. Or anything really.

I don't know how hard or if even possible to listen to these types of events on other OS's. I can definitely open an enhancement request though. 

Thanks for responding!
